### DISCUSSION TITLE
Using sqlalchemy MappedAsDataclass as Model doesn't work for init=False fields

### DISCUSSION THREAD

**Discussion by robertofalk** [pydantic/pydantic#5435] [Question]
Hi,

I'm trying to use the dataclass coming from sqlalchemy as the definition of my model, to be able to validate FastAPI request payload as follows:

```python
from sqlalchemy import Identity
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import MappedAsDataclass
from fastapi import FastAPI
from fastapi.testclient import TestClient


class Base(MappedAsDataclass, DeclarativeBase):
    pass

class User(Base):
    __tablename__ = "user_account"

    id: Mapped[int] = mapped_column(init=False, primary_key=True, server_default=Identity(always=True, start=1, increment=1, minvalue=1, cycle=False))
    name: Mapped[str]

app = FastAPI()

@app.post("/user")
def create_user(user: User):
    return True

client = TestClient(app)

def test_add_user():
    response = client.post("/user", json={"name": "john"})
    assert response.status_code == 200
``` 

When I run the test, the `response`  variable returns the following error:
```json
{"detail":[{"loc":["body","id"],"msg":"field required","type":"value_error.missing"}]}
```

The problem is, ID is an identity column, so I don't need it in the constructor, that's why I set the `init=False`. If I remove it, then the error changes:
```json
{"detail":[{"loc":["body"],"msg":"__init__() missing 1 required positional argument: 'id'","type":"type_error"}]}
```

It makes sense, now my payload is incomplete, since ID is missing, but the thing is, I'm trying to create a new user, so it's expected that ID will be empty. What am I missing here? How to use `identity` IDs together with pydantic?

Thanks!

**Comment by grumBit** [comment]
Hi @robertofalk,  I have the same question.

I'm trying to combine a pydantic model and SQLAlchemy ORM into a single class. My first step was trying to add in SQLAlchemy's `MappedAsDataclass`, however, I got the same problem with id as you. From there, I want to work out how to add in Pydantic's `Field` and `validator`, but I don't see a way to do this yet.

I started with this Pydantic model, with `Field`s and a `validator`;

```python
from pydantic import BaseModel, Field, validator

class VisitorNumbers(BaseModel):
    """
    Visitor numbers for a given day
    """
    start_date: date = Field(description="Start date")
    end_date: date = Field(description="End date")
    count: Optional[int] = Field(description="Number of visitors")
    note: Optional[str] = Field(description="Comments relating to the number of visitors on the day")

    @validator("end_date")
    def check_end_date(cls, end_date: date, values: Dict[Any, Any], **kwargs: Dict[Any, Any]) -> date:
        if end_date < values["start_date"]:
            raise ValueError("End date must be greater than or equal to start date")
        return end_date

    class Config:
        orm_mode = True
```

And this SQLAlchemy ORM;

```python
from datetime import date
from typing import Optional

from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship

class Base(DeclarativeBase):
    pass


class VisitorNumbers(Base):
    __tablename__ = "visitor_numbers"
    id: Mapped[int] = mapped_column(primary_key=True)
    start_date: Mapped[date]
    end_date: Mapped[date]
    count: Mapped[Optional[int]]
    note: Mapped[Optional[str]]
```

I got the same problem with the `id` as you did when I added in SQLAlchemy's `MappedAsDataclass` like;

```python
from datetime import date
from typing import Optional

from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship

class Base(MappedAsDataclass,  DeclarativeBase, dataclass_callable=pydantic.dataclasses.dataclass):
    pass


class VisitorNumbers(Base):
    __tablename__ = "visitor_numbers"
    id: Mapped[int] = mapped_column(primary_key=True)
    start_date: Mapped[date]
    end_date: Mapped[date]
    count: Mapped[int]
    note: Mapped[Optional[str]]
```

**Comment by grumBit** [comment]
Hi @robertofalk, I did some more work on this and got _something_ to work, but I don't like what I got going. Here are my thoughts;

The code below is the outcome of trying to combine pydantic and sqlalchemy such that the pydantic API and the sqlalchemy model are one and the same, while alembic can still generate migrations. The aim was to reduce potential errors where the API didn't include all the fields in the ORM.

- The concept of combining the API and ORM means everything in the ORM is always exposed in the API. That maybe be ok for now, but down the track with a more complex data model, we might want to change this, which would lead to rework. Basically, getting this to work may not be a good idea in the first place.

- I couldn't find a way to get Alembic to work when subclassing off _either_ pydantic's `BaseModel` or SA's `DeclaritiveBase`, which means all helper things like `.json()` and `.from_orm()` aren't available out of the box. I'm guessing there are workarounds to achieve the same, but...

- We can't use pydantics `Field` , as this breaks alembic, so we have to play with python's `field`, which makes defining the pydantic models harder and it isn't the recommended way to use pydantic.

- Pydantic's API documentation using `.schema_json()` is broken as the metadata now includes SA stuff, which can't be serialised. We _could_ write our own json encoder to workaround this, but...

- We can't use SA's `Mapped`, `mapped_column` and `relationship` and instead need to work directly with the `metadata`, which makes defining the ORMs considerably harder. That's ok-ish on trivial models, but may become tricky if the ORM becomes more complex and it isn't the recommended way to use SA. I.e. further down the track, we might need to dig in the bowels of SA to get things to work.

- In the alembic notes, they caution ["The vast majority of user issues with Alembic centers on the topic of what kinds of changes autogenerate can and cannot detect reliably"](https://alembic.sqlalchemy.org/en/latest/autogenerate.html#what-does-autogenerate-detect-and-what-does-it-not-detect). I don't know if we'll create problems down the track for alembic if we go down this off-piste route, but...

- SA can't do type inference, so a type missmatch b/w pydantic and SA can occur. Splitting out the API and ORM into separate classes has the same problem, so not having the type inference isn't any worse, but it does reduce the hoped-for benefits
Basically, with what I managed to get going below, we wind up with substandard pydantic models AND substandard SA models. It's possible there is a way to do this better than I've managed so far, but I feel it's not worth investigating further. Also, it will most likely mean we're going off-piste, which can lead to issues down the track and lead to bigger problems than what we set out to mitigate.

Basically, with what I managed to get going below, we wind up with substandard pydantic models AND substandard SA models. It's possible there is a way to do this better than I've managed so far, but I feel it's not worth investigating further. Also, it will most likely mean we're going off-piste, which can lead to issues down the track and lead to bigger problems than what we set out to mitigate in the first place!

```python
from dataclasses import field
from datetime import date
from typing import Optional

from pydantic.dataclasses import dataclass

from sqlalchemy.orm import registry
from sqlalchemy import Column, Integer, String, Date


mapper_registry = registry()


@mapper_registry.mapped
@dataclass
class VisitorNumberCOMBO:
    """
    Visitor numbers for a given day
    """

    day: date = field(
        metadata={
            "description": "Date of visits in ISO 8601 'YYYY-MM-DD' format",
            "sa": Column(Date, primary_key=True),
        }
    )
    count: Optional[int] = field(
        metadata={
            "description": "Number of visitors",
            "sa": Column(Integer),
        }
    )
    note: str = field(
        metadata={
            "description": "Comments relating to the number of visitors on the day",
            "sa": Column(String, nullable=False),
        }
    )

    __tablename__ = "visitor_number_combo"

    __sa_dataclass_metadata_key__ = "sa"
```

**Comment by robertofalk** [comment]
Hi @grumBit ,

Thanks for trying it out. Yes, I also don't like so much, but in the end what I did was simply duplicating the original `User` class in a `UserModel` class right next to it:

```python
from pydantic import BaseModel

class UserModel(BaseModel):
    name: str
```

Not the most elegant code but at least I can convert from the `UserModel` instance received in the `/user` endpoint to `User` like this:
```python
user = User(**user_model.dict())
```

**Comment by antont** [comment]
I guess you are aware that SQLModel implements these things? It allows using same classes both to define the SQLAlchemy tables and the FastAPI types. Alembic works with it also.

It has it's issues, is pending maintenance, and might benefit from porting to SA 2.0 Mapped and dataclass things, but anyway it's a nice integration of SA and FastAPI. The limitations are easily overcome by using SA definitions directly, like kwargs to Relationship.

**Reply by grumBit** [reply]
Sounds interesting!
